algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c

```diff
@@ -0,0 +1,118 @@
+
+//Circular Linked List
+//insertion and Deletion at the beginning of the Linked List
+#include<stdio.h>
+#include<stdlib.h>
+
+//  Structure Definition
+
+// 1. Data of Integer type.
+// 2. Pointer to the Next Node .
+struct node {
+    int data ;
+    struct node *next ;
+} ;
+struct node *head = NULL ;
+//Use tail pointer to avoid traversing the Linked List while inserting of the new created node
+//because after every insertion of a new node we need to point make the last node point to the newly created Node 
+struct node *tail = NULL ;
+
+//Function to Print the Linked List
+void printList(){
+	
+  if(head == NULL)
+  {
+      printf("\nList is empty");
+      return;
+  }
+  else{
+  
+
+    printf("\nItem in the Linked List are :  ");
+    // Initialize the ptr(pointer) with head location .
+    struct node *ptr = head ;
+    do{
+       printf("%d ",ptr->data) ;  //Print the value of the Node
+       ptr = ptr->next ;         //Increment the ptr 
+    }while(ptr!=head);           //// While Loop until we encounter a Node pointer ptr not pointing to the same location as head .
+    
+    }
+}
+
+//Insertion at the Beginning of the Circular Linked List
+
+void insertBeg(int data){
+ //Creating a link
+ //Malloc is for space allocation .
+ struct node *new_node = (struct node*) malloc(sizeof(struct node)) ;
+ if(new_node==NULL){
+     printf("Cannot create the Node");
+ }
+ else{
+     
+     new_node->data=data;
+     if(head==NULL){
+     	//making the head and the tail pointer to point to the newly created node
+         head = new_node ;
+         tail = new_node;
+     }
+     else{
+        //Point the new_node pointer to the current head
+        new_node->next = head ;
+        head = new_node ;
+     }
+     //making the tail pointer to point to the new head
+    tail->next=head;
+     
+     
+ }
+ 
+}
+
+//Deletion at the Beginning of the Circular Linked List
+
+void deleteBeg(){
+  //NO Deletion when list is empty
+   if(head == NULL) {
+    printf("List is Empty , deletion not possible") ;
+    return;
+   }
+   //when both the head and the tail pointer points to the same location
+   //i.e there is only one node left in the list
+   else if(head == tail)
+  {
+      free(head);
+      free(tail);
```

Comment:
Here this will throw a `double free error` , it occurs when you use `free()` more than once on the same memory address.
So here both `head` and `tail` both of them point to the same memory address, when you use `free(head)` everything in that memory address is released and since `tail` is pointing to this **released** memory address if you use `free(tail)` it will crash and throw in the error.

suggested changes
```C
else if(head == tail){
    free(head);
    head = NULL;
    tail = NULL;
}
```